ssl/ech/ech_internal.c

```diff
@@ -160,4 +214,1260 @@ int ossl_ech_conn_init(SSL_CONNECTION *s, SSL_CTX *ctx,
+            || !PACKET_get_1(&pkt, &pi_tmp) /* sessid len */
+            || !PACKET_get_bytes(&pkt, &pp_tmp, pi_tmp) /* sessid */
+            || !PACKET_get_net_2(&pkt, &pi_tmp) /* ciphersuite */
+            || !PACKET_get_1(&pkt, &pi_tmp) /* compression */
+            || !PACKET_get_net_2(&pkt, &pi_tmp)) /* len(extensions) */
+            return 0;
+        while (PACKET_remaining(&pkt) > 0 && done < 1) {
+            if (!PACKET_get_net_2(&pkt, &etype)
+                || !PACKET_get_net_2(&pkt, &elen))
+                return 0;
+            if (etype == TLSEXT_TYPE_ech) {
+                if (elen != OSSL_ECH_SIGNAL_LEN
+                    || !PACKET_get_bytes(&pkt, &acp, elen))
+                    return 0;
+                memcpy(acbuf, acp, elen);
+                done++;
+            } else {
+                if (!PACKET_get_bytes(&pkt, &pp_tmp, elen))
+                    return 0;
+            }
+        }
+        return done;
+    }
+    return 0;
+}
+
+/*
+ * return the h/s hash from the connection or ServerHello
+ * rmd is the returned h/s hash
+ * shbuf is the ServerHello
+ * shlen is the length of the ServerHello
+ * return 1 for good, 0 for error
+ */
+static int ech_get_md_from_hs(SSL_CONNECTION *s, EVP_MD **rmd,
+                              const unsigned char *shbuf, const size_t shlen)
+{
+    int rv;
+    size_t extoffset = 0, echoffset = 0, cipheroffset = 0;
+    uint16_t echtype;
+    const SSL_CIPHER *c = NULL;
+    const unsigned char *cipherchars = NULL;
+    EVP_MD *md = NULL;
+
+    /* this branch works for the server */
+    md = (EVP_MD *)ssl_handshake_md(s);
+    if (md != NULL) {
+        *rmd = md;
+        return 1;
+    }
+    /*
+     * if we're a client we'll fallback to hash from the chosen ciphersuite
+     * that means ECH acceptance depends on no bidding down, but that's ok
+     */
+    rv = ossl_ech_get_sh_offsets(shbuf, shlen, &extoffset, &echoffset,
+                                 &echtype);
+    if (rv != 1 || extoffset < 3) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_R_ECH_REQUIRED);
+        return 0;
+    }
+    cipheroffset = extoffset - 3;
+    cipherchars = &shbuf[cipheroffset];
+    c = ssl_get_cipher_by_char(s, cipherchars, 0);
+    if (c == NULL /* fuzzer fix */
+        || (md = (EVP_MD *)ssl_md(s->ssl.ctx, c->algorithm2)) == NULL) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    *rmd = md;
+    return 1;
+}
+
+/*
+ * make up a buffer to use to reset transcript
+ * for_hrr is 1 if we've just seen HRR, 0 otherwise
+ * shbuf is the output buffer
+ * shlen is the length of that buffer
+ * tbuf is the output buffer
+ * tlen is the length of that buffer
+ * chend returns the offset of the end of the last CH in the buffer
+ * fixedshbuf_len returns the fixed up length of the SH
+ * return 1 for good, 0 otherwise
+ */
+int ossl_ech_make_transcript_buffer(SSL_CONNECTION *s, int for_hrr,
```

Comment:
This function looks like it would benefit from being broken up into smaller chunks to make it more understandable. There is a lot of writing to packets and copying buffers around which looks very inefficient.

Reply:
fair enough - I'll take that as another input for transcript re-factoring in the next PR if that's ok?